### PR TITLE
feat(memory,query): counter correction v2 / chunked rate, increase, delta

### PR DESCRIPTION
**Pull request by velvia**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [x] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** (link exiting issues here : https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)

The `rate` and `increase` functions both rely on counter correction of a time series, and this is done using an iterator that goes over every sample of each time series queried.  This is slow and unnecessary for most time series since most counters when queries will not have a reset.

**New behavior :**

- Chunks (specifically the new `DoubleCounterAppender`) detect if there are any drops in incoming values, and if there are, sets a bit in the chunk header to indicate there are drops
- A new `RangeFunction` class keeps track of the amount of counter correction needed by each read chunk.  Corrections are carried over to successive chunks and increased if drops/resets are detected in successive chunks.  Special new methods in vector readers report corrected values and adjust corrections appropriately.
- A chunk-based rate and increase function reads just the values at the beginning and end of each window to compute rate.  This reduces running time of rate queries from O(n) to O(numWindows) roughly.
- Drop/reset detection for DoubleColumns is enabled using dataset type param `counter=true`
- multiple tests added to ensure rate calculations behave the same as old code

`IngestionBenchmark` does not show any drop in throughput (0.025 vs 0.025).

The two Query* Benchmarks show between roughly 2.5x and 4x improvement for `sum(rate(...))` queries.

This work is a prerequisite for increasing/Prom-style histograms work.

**BREAKING CHANGES**

`counter=true` needs to be added to datasets for rate() to keep working reliably.

### CODE REVIEW

**Comment by velvia** [comment]
@vishramachandran  suggestion is to go through commits one by one

File: memory/src/main/scala/filodb.memory/format/BinaryVector.scala [hunk 1]
```diff
@@ -448,6 +486,11 @@ object PrimitiveVectorReader {
   final def bitShift(vector: BinaryVectorPtr): Int = UnsafeUtils.getByte(vector + OffsetBitShift) & 0x03f
   final def signed(vector: BinaryVectorPtr): Boolean =
     (UnsafeUtils.getByte(vector + OffsetNBits) & SignMask) != 0
+  final def resetted(vector: BinaryVectorPtr): Boolean =
+    (UnsafeUtils.getShort(vector + OffsetNBits) & ResetMask) != 0
+
+  final def markReset(vector: BinaryVectorPtr): Unit =
```
Comment (broneill) [inline memory/src/main/scala/filodb.memory/format/BinaryVector.scala]: Naming this markReset kept tripping me up. Java streams and Buffers have a special use for "mark" and "reset". These are both verbs. How about setCounterReset and hasCounterReset? Not a big deal either way.

**Comment by vishramachandran** [comment]
@velvia A lot of changes in this PR related to counter query processing.
How did you validate that we did not regress?

We should add a bunch of tests that cover all counter related to various corner cases for counter correction and validate responses same before and after your PR. Were you able to do such an exercise ?

Earlier, I didn't add extensive test cases because the logic was similar to Prometheus'. But now that we are deviating, we need to validate correctness and consistency with Prom logic.

Here are ones that need to be included. Some of these have already been covered by you I think.

1. No counter correction
2. Counter correction at end of chunk and beginning of next chunk
3. Interpolation of rate
4. Fewer samples
5. Multiple corrections in one window
6. No changing in counter value for the entire chunk


Next question: Can bad input data or queries break the algorithm in a way that can create segvs ?

File: memory/src/main/scala/filodb.memory/format/BinaryVector.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (velvia) [reply memory/src/main/scala/filodb.memory/format/BinaryVector.scala]: I'll change everything to use "drop".

**Comment by velvia** [comment]
@vishramachandran I've addressed most feedback comments and added tests. 
Test cases 1 and 2 were already there.
Have added 4, 5, 6.

Re: "Interpolation of rate" ... what specifically would you like to see tested?

The calculation of rate remains exactly the same as before as far as the actual rate calculation code, including the interpolation logic, so I'm not sure we really need to test that.  The only thing that changes here is how the first and last values and timestamps are arrived at (via ChunkedWindowIterator instead of Sliding).  To test that, many tests have been added to RateFunctionSpec, which rely on the ChunkedWindowIterator and real raw RVs.

> Can bad input data or queries break the algorithm in a way that can create segvs ?

It should not.  The vector format has not changed at all, except for an unused bit that is now used to mark any drops in the data.  The reader which corrects uses on-heap memory to store corrected values.  There is no degradation in encoding due to non increasing or random data.

**Comment by velvia** [comment]
@vishramachandran  also surprised you didn't bring up that I only added chunked `rate` and `increase` but not `delta`.    Delta could be added as well, since the extrapolatedRate method is common to both sliding + chunked calculation, in theory it should produce identical results.

**Comment by vishramachandran** [comment]
@velvia Regarding delta function, it is only applied to gauges. Should never be applied to counters. This is per prometheus documentation.
